### PR TITLE
Split viewer-prototype in packages

### PR DESCRIPTION
**Pull request by tahini**
In prevision of the development of a vscode extension, the code in the
viewer-prototype has been mostly moved to 2 new packages:

* @trace-viewer/base: Queries the trace server for traces, experiments
and possible outputs
* @trace-viewer/react-components: Contains the various types of
visualization for trace data, queries the trace server for output data
to display

Those 2 packages can eventually be published to npm.

In the viewer-prototype directory only remains the theia specific module
contributions. So the application still works as before using theia,
it's just that most of the classes used by the trace-viewer have been
moved to new npm packages.

Signed-off-by: Geneviève Bastien <gbastien@versatic.net>

### CODE REVIEW

**Comment by tahini** [comment]
This PR is replacing PR #124. This one does not change our way of working right now. The theia browser and electron apps still work, but a good part of the code is moved to 2 packages that will be published on npm and consumed by the vscode and theia extension.

Thus this is a smaller bite to swallow on our way to a vscode extension.

**Comment by marcdumais-work** [comment]
I have not opened this repo in `Gitpod` for a while, so I do not know if this is new, but I notice the readability of the text is poor, in the example Electron application, accessed through vnc:

image

I found a VNC setting that fixes it: `scaling mode: remote resizing` instead of `local scaling`
image

image

**Comment by marcdumais-work** [comment]
When I briefly tested, I could get almost no graphs to display - maybe a single line for CPU usage view. Could be related to this backend trace, or not. Starting fresh in another container, that time I seem to crash it playing with the trace views. Not sure what happened exactly but the backend became unreachable. Maybe just bad luck. But I see kind of crazy amount of CPU being used, with one trace open and only the non-working CPU view. Like half to three quarters of the CPU power of the 16 cores of that host. Likely our container was killed because of too much resource usage.

Like before, I do not know if this is new or the same would happen on master.

```
root WARN Provided WebGL context does not have a stencil buffer, masks may not render correctly
root INFO 
    PixiJS 5.3.3 - ✰ WebGL 2 ✰      http://www.pixijs.com/    ♥♥♥ 


root WARN Provided WebGL context 
```

React is apparently not happy about something. Not sure if important:

```
root WARN Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-unsafe-component-lifecycles for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
* Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.

Please update the following components: ReactGridLayout, ResponsiveReactGridLayout
```

**Comment by tahini** [comment]
@marcdumais-work Thanks for trying it. You're talking about gitpod only? Or did you try it somewhere else more local to you? I'll investigate, try to see if this is new from this PR or if it was there now in master.

**Review by marcdumais-work — APPROVED** [review]
LGTM, thanks @tahini . I will let others review the TypeScript code changes, but the split and configs look good. I briefly tried on Gitpod and there is high CPU usage and other issues, but apparently already in master.

**Comment by marcdumais-work** [comment]
It will be very nice when we can use the browser version of your app, in Gitpod. All these GPU problems go away then.

**Comment by marcdumais-work** [comment]
> You're talking about gitpod only?

Correct. Most likely it will be better locally, but I'll let others confirm for now

**Review by bhufmann — CHANGES_REQUESTED** [review]
All the features work fine. Only comments tooltip the tooltips. Please let me know what you think for that.

**Comment by tahini** [comment]
@bhufmann I'm waiting for your feedback on this patch. 

Would you rather 
1- I do something custom to keep the tooltip for now, or 
2- wait for a later patch where we can make the react-components dispatch <stuff>, including tooltip, to <some listeners>?

**Comment by bhufmann** [comment]
> @bhufmann I'm waiting for your feedback on this patch.
> 
> Would you rather
> 1- I do something custom to keep the tooltip for now, or
> 2- wait for a later patch where we can make the react-components dispatch , including tooltip, to ?

@tahini, #170 implements a tooltip pop-up window and it doesn't need the signal (I think). So, I don't think we need a 1). Later I will see that we need for other purposes a way to communicate between the VSCode extension and the react-components.

File: viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx
Comment (bhufmann) [reply viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx]: Thanks

**Review by bhufmann — APPROVED** [review]
Looks good to me. Before merging I'd like to coordinate with the team for this.

**Comment by tahini** [comment]
Let me know when I can merge this, I just rebased on latest master, including the experiment sync.

**Comment by bhufmann** [comment]
> Let me know when I can merge this, I just rebased on latest master, including the experiment sync.

@tahini #170 won't be ready for the MVP. So, this split patch would remove the tooltip window that shows the state tooltip. I think we should have the original tooltip behaviour for the MVP. So, we have 2 options:
1) don't merge this patch before the MVP
2) update this patch to bring back the original tooltip functionality. 

Option 1) is the safe way for the MVP. However, option 2) would ok too, if the change doesn't take too long to implement. For option 2) we would have a minor update on our internal integration build.

What do you think?

**Comment by tahini** [comment]
I'll update this patch and bring back tooltip. I'd rather have this in sooner, it's a pain to rebase, especially with features that involve interactions between trace explorer and trace viewer. 

It shouldn't impact your internal MVP too much, especially that the packages are still in the same repo, no?

**Comment by bhufmann** [comment]
> I'll update this patch and bring back tooltip. I'd rather have this in sooner, it's a pain to rebase, especially with features that involve interactions between trace explorer and trace viewer.
> 
> It shouldn't impact your internal MVP too much, especially that the packages are still in the same repo, no?

The only impact I think is to have both the base and react-component packages published in our registry and reference them in our package.json. Should be straight forward.

Thanks for bringing the tooltip back in this PR.

**Comment by tahini** [comment]
The tooltips are back

File: README.md
Comment (bhufmann) [reply README.md:112]: This should be put back, right?

**Review by bhufmann — CHANGES_REQUESTED** [review]
It works fine. Please keep the tooltip in the README.md as well

**Review by bhufmann — APPROVED** [review]
Looks good to me. Thanks for this restructuring patch.

**Comment by bhufmann** [comment]
@tahini, patch is good to be merged

**Comment by tahini** [comment]
Thanks!

**Pull request merged**
